impl/monitor/monitor.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package monitor implements the monitor service. A monitor repeatedly polls a
+// key-transparency server's Mutations API and signs Map Roots if it could
+// reconstruct
+// clients can query.
+package monitor
+
+import (
+	"bytes"
+	"errors"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	"github.com/google/trillian"
+
+	tcrypto "github.com/google/trillian/crypto"
+
+	cmon "github.com/google/keytransparency/core/monitor"
+	ktpb "github.com/google/keytransparency/core/proto/keytransparency_v1_types"
+	mopb "github.com/google/keytransparency/core/proto/monitor_v1_types"
+
+	mupb "github.com/google/keytransparency/impl/proto/mutation_v1_service"
+)
+
+// Each page contains pageSize profiles. Each profile contains multiple
+// keys. Assuming 2 keys per profile (each of size 2048-bit), a page of
+// size 16 will contain about 8KB of data.
+const pageSize = 16
+
+var (
+	// ErrNothingProcessed occurs when the monitor did not process any mutations /
+	// smrs yet.
+	ErrNothingProcessed = errors.New("did not process any mutations yet")
+)
+
+// Server holds internal state for the monitor server.
+type Server struct {
+	client     mupb.MutationServiceClient
+	pollPeriod time.Duration
+
+	monitor        *cmon.Monitor
+	signer         *tcrypto.Signer
+	proccessedSMRs []*mopb.GetMonitoringResponse
+}
+
+// New creates a new instance of the monitor server.
+func New(cli mupb.MutationServiceClient,
+	signer *tcrypto.Signer,
+	logTree, mapTree *trillian.Tree,
+	poll time.Duration) *Server {
+	return &Server{
+		client:     cli,
+		pollPeriod: poll,
+		// TODO(ismail) use domain info to properly init. the monitor:
+		monitor:        &cmon.Monitor{},
+		signer:         signer,
+		proccessedSMRs: make([]*mopb.GetMonitoringResponse, 256),
```

Comment:
Please move non-grpc functions to their own object. 
They can be in the same package.